쥬스메이커 [STEP 3] Grumpy, Cathy

grumpy-sw opened the pull request:
안녕하세요 @hyunable !
STEP 3 PR입니다! 날씨가 많이 따뜻해졌네요~

---

고민했던 점


1.  뷰 컨트롤러 사이의 데이터 전송 
프로젝트 요구사항에 따르면 두 개의 뷰 컨트롤러 사이에 서로 데이터 전달이 필요한데, Delegate를 이용한 방식으로 구현하려 했으나 양방향 데이터 전달이 안되는 것 같습니다. 결국 쥬스 주문 화면에서 재고 관리 화면으로 현재 재고정보를 전달할 때 뷰의 변수에 직접 데이터를 전달하고, 수정된 재고정보를 전달할 때는 Delegate protocol에서 지정한 함수를 통해 데이터를 송신했습니다. 둘의 전달 방식이 달라서 조금 어색하다고 느껴지는데, 개선할 여지가 있을지 궁금합니다.

2. 변경된 재고를 화면에 띄우는 방법
JuiceOrderViewController 입장에서는 어떤 과일의 재고가 변경될 지 모르기 때문에 재고를 표시하는 모든 Label을 갱신해주는 것이 가장 안전하지만, 한 종류의 과일의 재고만 변경되거나 심지어 변경사항이 없더라도 전체 Label을 모두 갱신하기 위해 모든 값을 가져와야 합니다. 
이것이 비효율적이라는 생각이 들어 재고 수량의 변화가 일어난 과일들만 데이터로 전달하도록 했는데, 이 과정 때문에 '처음에 전달받은 과일의 수량과 현재 과일 수량이 동일한지' 확인하는 절차가 각 과일별로 1번씩 총 5번이 이루어지고 있습니다. 이 부분을 개선할 방법이 없을지 조언을 듣고 싶습니다!

Commit:
feat: 재고 수정 화면에 네비게이션바 추가

grumpy-sw committed:
feat: UIStepper Outlet&Action 추가

Commit:
feat: view controller간에 데이터 송/수신 구현

Commit:
Revert "feat: view controller간에 데이터 송/수신 구현"

This reverts commit 00b59dfd90520f357bbb73e9d55d67a8f42e62c8.

Commit:
feat: view controller간에 데이터 송/수신 구현

Commit:
feat: view controller 간에 데이터 송/수신 구현

grumpy-sw committed:
feat: 과일의 재고를 변경하는 함수 생성

FruitStore 내 updateStock(fruit: , amount:) 함수 구현

Commit:
feat: JuiceOrderViewController에서 과일 재고 반환하는 함수 생성, 과일 재고 데이터 전달

grumpy-sw committed:
feat: 전달받은 과일 재고 Label의 text에 출력, Outlet Collection 추가

Commit:
feat: UIStepper에서 변경한 과일 재고 데이터 전달

grumpy-sw committed:
feat: 재고 변경된 데이터를 juiceMaker에 반영 및 JuiceOrderViewController의 Label에 반영

Commit:
refactor: UIStepper collection으로 변경, 재고 변경 시에만 반영되도록 로직 변경, 최소값 설정

grumpy-sw committed:
refactor: 전반적인 refactoring 진행

코드 순서 수정
네이밍 수정
사용하지 않는 increaseStock함수, stockLabels 제거
가독성을 위해 setupFruitStockStepper함수 분리

grumpy-sw committed:
feat: AutoLayout 적용

StockInventoryView StackView로 Item 분할, Constraints 추가해서 간격 조정

Commit:
refactor: 메서드들을 private으로 수정

Commit:
Update README.md

grumpy-sw committed:
docs: README.md 오타 수정

hyunable commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -1,14 +1,12 @@
 
-class JuiceOrderViewController: UIViewController {
-    
+protocol dataDelegate {
```
protocol은 type명이기 때문에 대문자로 시작해주세요~!
그리고 보통 특정 클래스의 delegate라면 그 클래스명으로 시작하는 이름을 짓는것이 일반적입니다 : D

tip: 네이밍이 짓는 기준이 애매할때는 오리지널 컨포넌트들의 네이밍을 보면서 컨셉을 잡으면 더 좋아요 👍 

ex, UITableViewDelegate

grumpy-sw replied:
Naming convention을 찾아보고 JuiceOrderViewControllerDelegate 로 변경했습니다.

hyunable commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -1,14 +1,12 @@
-class JuiceOrderViewController: UIViewController {
-    
+protocol dataDelegate {
+    func sendData(fruits: [FruitType: Int])
```
메서드 명도 UIKit쪽 컴포넌트들 네이밍 스타일을 참고해서 변경해볼까요? 

ex, func tableView(UITableView, willDisplay: UITableViewCell, forRowAt: IndexPath)

grumpy-sw replied:
쥬스 주문 뷰에 재고의 변화가 있다는 것을 알린다는 의미를 담아 ```juiceOrderViewControllerHasChanges()```라고 변경했습니다. 프로토콜을 채택한 뷰 컨트롤러로부터 위임받은 메서드라고 봐야 할까요? 어제 말씀드린 것처럼 행위의 주체와 대상이 헷갈려서 이름짓기가 힘든 것 같네요..ㅎㅎ

hyunable commented on JuiceMaker/JuiceMaker/Controller/StockInventoryViewController.swift:
```diff
@@ -1,11 +1,105 @@
 class StockInventoryViewController: UIViewController {
+    var fruitStockStatus: [FruitType: Int] = [:]
+    let minimumNumberOfStock: Double = 0
+    var dataDelegate: dataDelegate?
```
delegate는 약한 참조로 선언해주세요~!

grumpy-sw replied:
변경했습니다!

hyunable reviewed (commented):
1. FruiteStore는 저희가 class로 만들어주었죠!  왜 class일까요? 
만약 `fruitStockStatus`메서드를 통해 과일데이터를 넘겨주는 것 대신 FruitStore 자체를 넘겨주게 되면 어떻게될까요?  StockInventoryViewController에서 FruitStore의 값을 변경해주면 JuiceOrderViewController에서 들고있는 FruitStore의 값은 어떻게 변경될까요?  이 관점에서 데이터를 변경하는 방식을 한번 변경해볼까요? (이해가안가는부분이 있다면 바로 코멘트나 디코로 막히거나 이해안되는 부분을 말씀해주세요~!)


2. 변경된 부분만 업데이트를 해주고 싶으셨군요 ㅎㅎ 접근 너무 좋은것 같아요. 하지만 제 생각에는 어떤 파트가 변경되었고 그부분만 변경해줘 하는 시나리오가 좀 더 비용이 많이 든다고생각해요~! 라벨 셋팅전에 모든 라벨이 validation체크를 해야하기때문이죠. 그래서 전체 라벨을 화면이 appear될때마다 업데이트해주는게 어떨까 싶어요 ㅎㅎ

Commit:
refactor: 프로토콜과 프로토콜 메서드 naming 수정, delegate를 약한 참조로 선언

grumpy-sw committed:
feat: 변경된 과일만 데이터로 전송하는 부분 제거

Commit:
refactor: 과일 데이터 전달하는 로직 변경

fruitStockStatus 메서드를 사용하지 않고 FruitStore를 전달해서 재고 데이터 변경

grumpy-sw committed:
feat: 커스텀 클래스 JuiceButton, FruitLabel 추가

Commit:
refactor: FruitLabel custom class를 사용하여 코드 수정

grumpy-sw committed:
feat: 커스텀 클래스 FruitStepper 추가, Stepper 누를때마다 increase/decrease하도록 수정

grumpy-sw committed:
rename: 커스텀 클래스 및 프로토콜을 독립된 파일로 분리

Commit:
rename: JuiceMakerError 파일을 enum 폴더로 이동

grumpy-sw committed:
chore: 대소문자 잘못 입력되어있는 오타 수정

JuiceOrderViewControllerHasChanges() -> juiceOrderViewControllerHasChanges() 로 변경

hyunable reviewed (approved):
다들 고생 많으셨어요 :) 
리뷰내용 잘 반영해주셔서 감사합니다~!
